### DISCUSSION TITLE
Import Module for Script

### DISCUSSION THREAD

**Discussion by CassidyHunt** [netbox-community/netbox#9339] [Q&A]
I am sure this has been asked before and I apologize if it is a repeat.  I couldn't find it in the history or on google.  I am new to Netbox and have been building custom scripts.  My Python runs just fine at the command line and does what it is suppose to.  I save the file in the scripts folder, with the proper items found in the documentation, but I get ModuleNotFound Error.  I have removed the import statement and script page pulls up fine.  So what am I missing to be able to use those modules in the script section of Netbox?

Thank you in advance for the help.

Cassidy

**Comment by jeremystretch** [comment]
Are you able to share your script, or a portion of it, for someone else to test? It's difficult to guess at what the issue might be.

**Comment by CassidyHunt** [comment]
Sure.  Really it is anything that isn't already baked in.  Here is an example:

```
import ping3
from extras.scripts import *

class PingIpAddress(Script):
    class Meta():
        name = 'Ping an IP Address'
        description = "This utility will accept an IP Address for input and ping the address"
        commit_default = False
        field_order = ['ipAddress']

    ipAddress = StringVar(
        description = "IP Address To Ping"
    )

    def run(self, data, commit):
        output = ping(data['ipAddress'])
        return str(output)
```

This causes 
```
<class 'ModuleNotFoundError'>

No module named 'ping3'

Python version: 3.8.10
NetBox version: 3.1.9
```

However this code works because the modules are baked in to the install:

```
import subprocess
import ipaddress
from extras.scripts import *

class PingSubnet(Script):
    class Meta():
        name = 'Ping an entire subnet'
        description = "This utility will accept a subnet and ping all the hosts"
        commit_default = False
        field_order = ['Subnet']

    Subnet = StringVar(
        description = "Enter Subnet in CIDR format (ex.192.168.1.0/24))"
    )

    def run(self, data, commit):

        ipNetwork = ipaddress.ip_network(data['Subnet'])

        allHosts = list(ipNetwork.hosts())

        results = ''

        for i in range(len(allHosts)):
            output = subprocess.run(["ping","-c","1",str(allHosts[i])], stdout=subprocess.PIPE)
            if output.returncode == 0:
                results = results + str(allHosts[i]) + " Successful!\n"
            else:
                results = results + str(allHosts[i]) + " Failed!\n"

        return results

class PingIPAddress(Script):
    class Meta():
        name = 'Ping an IP Address'
        description = "This utility will accept an IP Address for input and ping that address"
        commit_default = False
        field_order = ['ipAddress']

    ipAddress = StringVar(
        description = "IP Address To Ping"
    )

    def run(self, data, commit):
        output = subprocess.run(["ping","-c","1",data['ipAddress']], stdout=subprocess.PIPE)
        return str(output) + " " + output.stdout.decode()
```

They achieve the same goal just different ways.  However they illustrate my problem.  Why can't I install ping3 or for that matter any other module and use it just as I have the ones baked in.  I can use the ping3 directly from the command line after installing it using pip.  I just can't use it from netbox when I drop it in the scripts directory.  I get that Module not found issue.

Hope this helps figure out the solution.

**Comment by CassidyHunt** [comment]
Sorry guess my quotes messed up on the post so please excuse the abrupt exit code enter code boxes.

**Reply by candlerb** [reply]
You can edit your post. Put three backticks (` ``` `) on a line of their own, before and after the code block.

**Comment by bogdancordos** [comment]
I think you only need to add "ping3" to your local_requirements.txt file within your NetBox installation and run ./upgrade.sh
or install ping3 into the NetBox virtual environment.
It works on your CLI because ping3 might be installed on the global environment and NetBox can't access it.

**Reply by CassidyHunt** [reply]
I will give it a try and let you know how I come out.
